drivers/mpu9250/MPU9250_mag.hpp

```diff
@@ -0,0 +1,178 @@
+
+// MPU9250 Magnetometer Register Addresses: Defines only the register addresses
+// used in MPU9250 driver.
+#define MPU9250_MAG_REG_WIA		0x00
+#define MPU9250_MAG_REG_ST1		0x02
+#define MPU9250_MAG_REG_DATA	0x03
+#define MPU9250_MAG_REG_HXL     0x03
+#define MPU9250_MAG_REG_ST2		0x09
+#define MPU9250_MAG_REG_CNTL1	0x0a
+#define MPU9250_MAG_REG_ASAX	0x10
+#define MPU9250_MAG_REG_ASAY    0x11
+#define MPU9250_MAG_REG_ASAZ    0x12
+
+// Bit definitions for the magnetometer registers
+#define BIT_MAG_CNTL1_MODE_POWER_DOWN 				 0x0
+#define BIT_MAG_CNTL1_MODE_SINGLE_MEASURE_MODE 		 0x1
+#define BIT_MAG_CNTL1_MODE_CONTINUOUS_MEASURE_MODE_1 0x2
+#define BIT_MAG_CNTL1_MODE_CONTINUOUS_MEASURE_MODE_2 0x6
+#define BIT_MAG_CNTL1_FUSE_ROM_ACCESS_MODE 		 	 0xF
+#define BIT_MAG_CNTL1_16_BITS 						 0x10
+#define BIT_MAG_HOFL                                 0x08
+
+#define MAG_ERROR_DATA_OVERFLOW                      -3
+
+enum mag_sample_rate_e
+{
+	MPU9250_MAG_SAMPLE_RATE_8HZ = 0,
+	MPU9250_MAG_SAMPLE_RATE_100HZ = 1,
+	NUM_MPU9250_MAG_SAMPLE_RATES
+};
+
+class MPU9250_mag
+{
+public:
+	MPU9250_mag(MPU9250 &imu, enum mag_sample_rate_e sample_rate) :
+			_mag_initialized(false), _sample_rate(sample_rate), _imu(imu)
+	{
+	};
+
+	/// @brief
+	/// Called to initialize the magnetometer connection via the
+	/// internal I2C bus of the sensor.  The gyro and accelerometer
+	/// must have been previously configured.
+	/// @return
+	/// - 0 on success,
+	/// - -errno on failure
+	int initialize(int output_data_rate_in_hz);
+
+	/// @brief
+	/// Reads the sensitivity values contained in the FUSE memory of the mag and
+	/// generates values used internally to process mag measurements.
+	/// @return
+	/// - 0 on success, -
+	/// errno on failure
+	int get_sensitivity_adjustment(void);
+
+	/// @brief
+	/// Verifies the presence of the mag on the internal I2C bus, by querying
+	/// for the known device ID.
+	/// @return
+	/// - 0 on success
+	/// - -errno on failure
+	int detect(void);
+
+	/// @brief
+	/// Writes a value to the specified IMU register, and verifies that it was
+	/// successfully written by attempting to read it back.
+	/// @return
+	/// - 0 on success
+	/// - -errno on failure
+	int write_imu_reg_verified(int reg, uint8_t val, uint8_t mask);
+
+	/// @brief
+	/// Reads the value of the specified magnetometer register and returns the
+	/// register value in the "val" parameter.
+	/// @return
+	/// - 0 on success
+	/// - -errno on failure
+	int read_reg(uint8_t reg, uint8_t *val);
+
+	/// @brief
+	/// Writes the value passed in to the specified magnetometer register.
+	/// @return
+	/// - 0 on success
+	/// - -errno on failure
+	int write_reg(uint8_t reg, uint8_t val);
+
+	/// @brief
+	/// Process the data passed in to generate mag values in Gauss units.
+	/// The values passed in are modified and returned in the same parameter.
+	/// @return
+	/// - 0 on success
+	/// - -errno on failure
+	int process(struct mag_data &data);
+
+protected:
```

Comment:
To allow access by derived classes, but it could be private since such extension is unlikely.